Spigot-Server-Patches/0333-Add-entity-count-cache.patch

```diff
@@ -0,0 +1,221 @@
+From 5b101db469123c2f92ff82fc04a8b77274296bcf Mon Sep 17 00:00:00 2001
+From: Colin Godsey <crgodsey@gmail.com>
+Date: Wed, 8 Aug 2018 10:10:06 -0600
+Subject: [PATCH] Add entity count cache
+
+
+diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
+index d6b56d685..3d973deb7 100644
+--- a/src/main/java/net/minecraft/server/MinecraftServer.java
++++ b/src/main/java/net/minecraft/server/MinecraftServer.java
+@@ -1034,6 +1034,14 @@ public abstract class MinecraftServer implements IAsyncTaskHandler, IMojangStati
+                 }
+                 // CraftBukkit end */
+ 
++                /*
++                 * Paper - clear count cache every once in a while.
++                 * Cache can never be entirely trusted because
++                 * an entity may become persistent or not
++                 * during its lifetime.
++                 */
++                if (this.ticks % 600 == 0) worldserver.clearEntityCountCache();
++
+                 this.methodProfiler.a("tick");
+ 
+                 CrashReport crashreport;
+diff --git a/src/main/java/net/minecraft/server/SpawnerCreature.java b/src/main/java/net/minecraft/server/SpawnerCreature.java
+index f525fd1b4..494759a1c 100644
+--- a/src/main/java/net/minecraft/server/SpawnerCreature.java
++++ b/src/main/java/net/minecraft/server/SpawnerCreature.java
+@@ -113,7 +113,7 @@ public final class SpawnerCreature {
+                 // CraftBukkit end
+ 
+                 if ((!enumcreaturetype.c() || flag1) && (enumcreaturetype.c() || flag) && (!enumcreaturetype.d() || flag2)) {
+-                    k = worldserver.a(enumcreaturetype.a());
++                    k = worldserver.getCreatureCount(enumcreaturetype); // Paper - entity count cache
+                     int l1 = limit * i / b; // CraftBukkit - use per-world limits
+ 
+                     if (k <= l1) {
+diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
+index 004c3ec47..baf73a411 100644
+--- a/src/main/java/net/minecraft/server/World.java
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -1,28 +1,20 @@
+ 	package net.minecraft.server;
+ 
+-import co.aikar.timings.Timings;
```

Comment:
... thanks intellij

Reply:
You'll want to turn off auto optimizing imports while working on paper :)